### PR TITLE
feat:  remove mutability from the WasiCtx Table

### PR DESCRIPTION
**Pull request by haraldh**
This patch adds interior mutability to the WasiCtx Table and the Table elements.

Major pain points:
* `File` only needs `RwLock<cap_std::fs::File>` to implement `File::set_fdflags()` on Windows, because of https://github.com/bytecodealliance/system-interface/blob/da238e324e752033f315f09c082ad9ce35d42696/src/fs/fd_flags.rs#L210-L217
* Because `File` needs a `RwLock` and `RwLock*Guard` cannot be hold across an `.await`, The `async` from `async fn num_ready_bytes(&self)` had to be removed
* Because `File` needs a `RwLock` and `RwLock*Guard` cannot be dereferenced in `pollable`, the signature of `fn pollable(&self) -> Option<rustix::fd::BorrowedFd>` changed to `fn pollable(&self) -> Option<Arc<dyn AsFd + '_>>`

Related: https://github.com/bytecodealliance/wasmtime/issues/5235

### CODE REVIEW

**Comment by bjorn3** [comment]
This will result in a performance regression due to all these locking operations, right? Are you making this change to add multi-threading support?

File: crates/wasi-common/src/dir.rs [hunk 1]
```diff
@@ -98,67 +99,72 @@ pub trait WasiDir: Send + Sync {
...
         }
     }
-    pub fn drop_caps_to(&mut self, caps: DirCaps, file_caps: FileCaps) -> Result<(), Error> {
-        self.capable_of_dir(caps)?;
+    pub fn drop_caps_to(&self, dir_caps: DirCaps, file_caps: FileCaps) -> Result<(), Error> {
+        self.capable_of_dir(dir_caps)?;
         self.capable_of_file(file_caps)?;
-        self.caps = caps;
-        self.file_caps = file_caps;
+        *self.caps.write().unwrap() = DirFdStat {
```
Comment (bjorn3) [inline crates/wasi-common/src/dir.rs:154]: Race condition here. Another call to drop_caps_to could have dropped different permissions between the capable_of_{dir,file} call and this write, which would result in capabilities being added back despite the other drop_caps_to returning a success,

Comment (haraldh) [reply crates/wasi-common/src/dir.rs:154]: Thank for spotting this!

File: crates/wasi-common/src/table.rs [hunk 2]
```diff
@@ -10,83 +12,89 @@ use std::collections::HashMap;
...
-        self.map.remove(&key)
+    pub fn renumber(&self, from: u32, to: u32) -> Result<(), Error> {
+        let from_entry = self
+            .map
+            .write()
+            .unwrap()
+            .remove(&from)
+            .ok_or(Error::badf())?;
+        self.map.write().unwrap().insert(to, from_entry);
+        Ok(())
```
Comment (alexcrichton) [inline crates/wasi-common/src/table.rs:98]: This should probably be an atomic operation rather than two seaparate ones.

File: crates/wasi-common/src/table.rs [hunk 3]
```diff
@@ -10,83 +12,89 @@ use std::collections::HashMap;
...
+                .next_key
+                .fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+            if self.map.read().unwrap().contains_key(&key) {
                 continue;
             }
-            self.map.insert(key, a);
+            self.map.write().unwrap().insert(key, a);
             return Ok(key);
         }
     }
```
Comment (alexcrichton) [inline crates/wasi-common/src/table.rs:50]: This should be all one atomic operation which can be protected by one lock, there's no need to use both atomics and locks here.

File: crates/wasi-common/src/snapshots/preview_1.rs [hunk 4]
```diff
@@ -1016,36 +1010,34 @@ impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiCtx {
...
         buf: &GuestPtr<'a, u8>,
         buf_len: types::Size,
     ) -> Result<(), Error> {
         let mut buf = buf
             .as_array(buf_len)
             .as_slice_mut()?
             .expect("cannot use with shared memories; see https://github.com/bytecodealliance/wasmtime/issues/5235 (TODO)");
-        self.random.try_fill_bytes(buf.deref_mut())?;
+        self.random
+            .write()
```
Comment (alexcrichton) [inline crates/wasi-common/src/snapshots/preview_1.rs:1031]: I don't think there's any need to serialize multiple threads like this, so the randomness provider should move to a trait with `&self` rather than `&mut self`

File: crates/wiggle/generate/src/module_trait.rs [hunk 5]
```diff
@@ -82,9 +82,9 @@ pub fn define_module_trait(m: &Module, settings: &CodegenSettings) -> TokenStrea
         };
 
         if is_anonymous {
-            quote!(#asyncness fn #funcname(&mut self, #(#args),*) -> #result; )
+            quote!(#asyncness fn #funcname(&self, #(#args),*) -> #result; )
         } else {
-            quote!(#asyncness fn #funcname<#lifetime>(&mut self, #(#args),*) -> #result;)
+            quote!(#asyncness fn #funcname<#lifetime>(&self, #(#args),*) -> #result;)
         }
```
Comment (alexcrichton) [inline crates/wiggle/generate/src/module_trait.rs:88]: Wiggle is used for more than just WASI, and I don't think it's suitable to make this change. Providing `&mut` access shows how Wasmtime gives this ability even if WASI doesn't take advantage of it.

Instead the traits should still give access to `&mut self` but `WasiCtx` would internally hold `Arc<Table>` or something like that which is where the requirement for locking comes in (because `&mut Arc<T>` still only gives you `&T`)

File: crates/wasi-common/src/file.rs [hunk 6]
```diff
@@ -1,67 +1,73 @@
...
-    fn pollable(&self) -> Option<rustix::fd::BorrowedFd> {
+    fn pollable(&self) -> Option<Arc<dyn AsFd + '_>> {
         None
     }
 
     #[cfg(windows)]
-    fn pollable(&self) -> Option<io_extras::os::windows::RawHandleOrSocket> {
+    fn pollable(&self) -> Option<Arc<dyn AsRawHandleOrSocket + '_>> {
         None
     }
```
Comment (alexcrichton) [inline crates/wasi-common/src/file.rs:24]: How come these need changing? The prior signatures I think should work?

File: crates/wasi-common/src/dir.rs [hunk 7]
```diff
@@ -98,67 +100,83 @@ pub trait WasiDir: Send + Sync {
...
-            Err(err.context(format!("desired rights {:?}, has {:?}", caps, self.caps)))
+            Err(err.context(format!(
+                "desired rights {:?}, has {:?}",
+                caps, fdstat.dir_caps
+            )))
         }
     }
-    pub fn capable_of_file(&self, caps: FileCaps) -> Result<(), Error> {
-        if self.file_caps.contains(caps) {
+    fn locked_capable_of_file(
```
Comment (alexcrichton) [inline crates/wasi-common/src/dir.rs:148]: Could this perhaps be a method on `DirFdStat`?

File: crates/wasi-common/cap-std-sync/src/file.rs [hunk 8]
```diff
@@ -14,11 +15,48 @@ use wasi_common::{
...
+
+#[cfg(windows)]
+impl AsRawHandleOrSocket for BorrowedFile {
+    #[inline]
+    fn as_raw_handle_or_socket(&self) -> RawHandleOrSocket {
+        self.0.as_raw_handle_or_socket()
+    }
+}
+
+pub struct File(RwLock<cap_std::fs::File>);
```
Comment (alexcrichton) [inline crates/wasi-common/cap-std-sync/src/file.rs:51]: I think ideally this is avoided because a system file doesn't need an extra mutex around it since it should already be able to perform all necessary operations concurrently.

**Review by alexcrichton — COMMENTED** [review]
Have you been able to run any benchmarks to see the effect of this change?

**Comment by abrown** [comment]
@haraldh, you should now be able to run the applicable benchmarks with something like `cargo bench wasi/` from the top-level project directory.

File: crates/wasi-common/src/file.rs (hunk 6 above)
Comment (haraldh) [reply crates/wasi-common/src/file.rs:24]: In the method a `RwLockWriteGuard<BorrowedFd>` would be produced, which can't be dereferenced for return.

File: crates/wasi-common/src/dir.rs (hunk 7 above)
Comment (haraldh) [reply crates/wasi-common/src/dir.rs:148]: Yeah, sure!

File: crates/wasi-common/cap-std-sync/src/file.rs (hunk 8 above)
Comment (haraldh) [reply crates/wasi-common/cap-std-sync/src/file.rs:51]: Well, it's only `cap_std::fs::File::set_fd_flags(&mut self, set_fd_flags: SetFdFlags<Self>) -> io::Result<()>` which needs it.
```

Comment (haraldh) [reply crates/wasi-common/cap-std-sync/src/file.rs:51]: Maybe, we should start removing the `mut` there.

File: crates/wiggle/generate/src/module_trait.rs (hunk 5 above)
Comment (haraldh) [reply crates/wiggle/generate/src/module_trait.rs:88]: `Arc<T>` sounds like a plan. Thanks!

File: crates/wasi-common/src/table.rs (hunk 2 above)
Comment (haraldh) [reply crates/wasi-common/src/table.rs:98]: You are right! Thanks!

**Comment by haraldh** [comment]
> Have you been able to run any benchmarks to see the effect of this change?

I haven't seen any significant change (-2% at most), but the wasi benches have a relatively big difference from run to run anyway.

File: crates/wasi-common/src/ctx.rs [hunk 9]
```diff
@@ -85,17 +86,17 @@ impl WasiCtx {
...
     }
 
-    pub fn set_stdout(&mut self, mut f: Box<dyn WasiFile>) {
+    pub fn set_stdout(&self, mut f: Box<dyn WasiFile>) {
         let rights = Self::stdio_rights(&mut *f);
         self.insert_file(1, f, rights);
     }
 
-    pub fn set_stderr(&mut self, mut f: Box<dyn WasiFile>) {
+    pub fn set_stderr(&self, mut f: Box<dyn WasiFile>) {
```
[the remark this answers is not in the document]
Comment (haraldh) [reply crates/wasi-common/src/ctx.rs:99]: Yeah, sure, but it doesn't need the `mut` now anymore with the table having interior mutability

File: crates/wiggle/generate/src/module_trait.rs (hunk 5 above)
Comment (haraldh) [reply crates/wiggle/generate/src/module_trait.rs:88]: What do you think about:
```rust
impl wasi_snapshot_preview1::WasiSnapshotPreview1 for Arc<WasiCtx>
```

File: crates/wasi-common/cap-std-sync/src/file.rs (hunk 8 above)
Comment (haraldh) [reply crates/wasi-common/cap-std-sync/src/file.rs:51]: seems to be in crate `system-interface` ... `trait GetSetFdFlags`

Comment (haraldh) [reply crates/wasi-common/cap-std-sync/src/file.rs:51]: https://github.com/bytecodealliance/system-interface/blob/da238e324e752033f315f09c082ad9ce35d42696/src/fs/fd_flags.rs#L210-L217

File: crates/wiggle/generate/src/module_trait.rs (hunk 5 above)
Comment (alexcrichton) [reply crates/wiggle/generate/src/module_trait.rs:88]: I would personally prefer to put the `Arc` inside of the `WasiCtx` to make it more ergonomic to use but either way achieves the same goal.

File: crates/wasi-common/src/file.rs (hunk 6 above)
Comment (alexcrichton) [reply crates/wasi-common/src/file.rs:24]: Sorry let me restate. From a system API perspective a lock is not needed here, so I don't know where the rwlock is motivated from. I would expect that these signatures remain the same since this is intended for bottoming out in the underlying platform-specific interface which handles all synchronization internally in the kernel.

File: crates/wasi-common/cap-std-sync/src/file.rs (hunk 8 above)
Comment (alexcrichton) [reply crates/wasi-common/cap-std-sync/src/file.rs:51]: Maybe @sunfishcode could help here, I'm not familiar enough with any of the methods/interfaces here to know what to do off-hand. At a base level though I don't at least currently understand why an `RwLock` or other form of synchronization is required for I/O operations on system files.

File: crates/wasi-common/src/file.rs (hunk 6 above)
Comment (haraldh) [reply crates/wasi-common/src/file.rs:24]: It's only because of `File::set_fd_flags()` needing &mut self

File: crates/wasi-common/src/snapshots/preview_1.rs (hunk 4 above)
Comment (haraldh) [reply crates/wasi-common/src/snapshots/preview_1.rs]: deferred this to a later PR

File: crates/wiggle/generate/src/module_trait.rs (hunk 5 above)
Comment (haraldh) [reply crates/wiggle/generate/src/module_trait.rs]: deferred this for a later PR

File: crates/wasi-common/src/dir.rs (hunk 7 above)
Comment (haraldh) [reply crates/wasi-common/src/dir.rs]: done

File: crates/wasi-common/cap-std-sync/src/file.rs (hunk 8 above)
Comment (haraldh) [reply crates/wasi-common/cap-std-sync/src/file.rs:51]: @alexcrichton it's because Windows has to reopen the file with new Flags, rather than modifying the current handle. There is just no native function for this.

Comment (alexcrichton) [reply crates/wasi-common/cap-std-sync/src/file.rs:51]: Hm ok sorry I'm having trouble phrasing my question/concern apparently. Let me try again. I do not understand the relevance to the `#[cfg(windows)]` block you mentioned above. I don't know what that function is, why it exists, why it's Windows-specific, or why it's at all related to `wasi-common` and this PR. What in WASI supports reopening a file with new flag and modifying a handle?

I don't understand these connections so I'm under the impression that the rwlock shouldn't be needed here. Can you help me understand the connections here?

Comment (haraldh) [reply crates/wasi-common/cap-std-sync/src/file.rs:51]: `wasmtime/crates/wasi-common/cap-std-sync/src/file.src`

```rust
impl WasiFile for File {
    // ...
    async fn set_fdflags(&self, fdflags: FdFlags) -> Result<(), Error> {
        if fdflags.intersects(
            wasi_common::file::FdFlags::DSYNC
                | wasi_common::file::FdFlags::SYNC
                | wasi_common::file::FdFlags::RSYNC,
        ) {
            return Err(Error::invalid_argument().context("cannot set DSYNC, SYNC, or RSYNC flag"));
        }
        let mut file = self.0.write().unwrap();
        let set_fd_flags = (*file).new_set_fd_flags(to_sysif_fdflags(fdflags))?;
        (*file).set_fd_flags(set_fd_flags)?;
        Ok(())
    }
    // ...    
}
```

This is the only function which needs the `RwLock::write()` (`&mut self` before`) of the whole `impl WasiFile for File`. 
It calls out to `cap_std::fs::File::set_fd_flags()`, which wants `&mut self`.

`cap_std::fs::File` automatically implements the `GetSetFdFlags` of the `system-interface` crate, which provides the `set_fd_flags()` method, to which I linked above.

Apparently the only way to change the "fd" flags on windows, is to reopen the file, which will get you another Handle. So the way it is implemented, it changes the inner handle for the File handle.

```rust
        *self = set_fd_flags.reopened;
```

This is why `self` has to be mutable for `set_fd_flags`. And that is only the case on Windows.

Comment (alexcrichton) [reply crates/wasi-common/cap-std-sync/src/file.rs:51]: Ok that makes sense, thanks for explaning. Personally I feel like this has a significant impact enough that I think it should be solved before merging. 

Digging in further, this feels like a bit of an odd API. All of the `*SYNC` flags aren't supported on Unix but they are sort of supported on Windows so there's already precedent for platform-specific behavior, although I'm not sure if that's desired or not. Additionally Windows rejects the `NONBLOCK` flag which means the only thing Windows supports here is `APPEND`. Forcing the entire API to have a lock unnecessarily for all platforms just for this one use case feels a bit off to me.

I would propose a course of action going forward entailing:

* Rethinking this API for WASI and whether it makes sense (cc @sunfishcode), for example changing the API could remove the need for this implementation
* In the meantime, change the trait signature to `self: &Arc<Self>` or something like that and use `Arc::get_mut` to keep this working in single-threaded situations while otherwise just returning an error in multi-threaded situations.

I think that should remove the need for the `RwLock`/`poll` changes, set this up for having a real solution in the long run, and getting something reasonable working for now.

Comment (bstrie) [reply crates/wasi-common/cap-std-sync/src/file.rs:51]: I'm happy to adapt this PR as described above. I agree that it's better to have something reasonable that we can iterate on in the short term.

Comment (abrown) [reply crates/wasi-common/cap-std-sync/src/file.rs:51]: I took a look at @alexcrichton's suggestion yesterday and abandoned after fixing compiler errors for several hours. The switch to `Arc<Self>` affects many places and after fixing up many of these places I was not sure if this approach would even work in the end. Here is the WIP commit, but note that not all errors are fixed at this point (probably 20+ left): https://github.com/abrown/wasmtime/compare/pr-5326...abrown:wasmtime:pr-5326-arc.  @alexcrichton can you take a look?

Comment (alexcrichton) [reply crates/wasi-common/cap-std-sync/src/file.rs:51]: Ah sorry I meant moreso `self: &Arc<Self>` on just that one method. Trying that myself runs afoul of object safety issues, so the alternative is:

<details>

```diff
diff --git a/crates/wasi-common/cap-std-sync/src/file.rs b/crates/wasi-common/cap-std-sync/src/file.rs
index c54a5ead7..c184486a7 100644
--- a/crates/wasi-common/cap-std-sync/src/file.rs
+++ b/crates/wasi-common/cap-std-sync/src/file.rs
@@ -93,7 +93,7 @@ impl WasiFile for File {
         let fdflags = get_fd_flags(&*file)?;
         Ok(fdflags)
     }
-    async fn set_fdflags(&self, fdflags: FdFlags) -> Result<(), Error> {
+    async fn set_fdflags(&mut self, fdflags: FdFlags) -> Result<(), Error> {
         if fdflags.intersects(
             wasi_common::file::FdFlags::DSYNC
                 | wasi_common::file::FdFlags::SYNC
@@ -101,7 +101,7 @@ impl WasiFile for File {
         ) {
             return Err(Error::invalid_argument().context("cannot set DSYNC, SYNC, or RSYNC flag"));
         }
-        let mut file = self.0.write().unwrap();
+        let file = self.0.get_mut().unwrap();
         let set_fd_flags = (*file).new_set_fd_flags(to_sysif_fdflags(fdflags))?;
         (*file).set_fd_flags(set_fd_flags)?;
         Ok(())
diff --git a/crates/wasi-common/cap-std-sync/src/net.rs b/crates/wasi-common/cap-std-sync/src/net.rs
index b46d8d7da..921622d68 100644
--- a/crates/wasi-common/cap-std-sync/src/net.rs
+++ b/crates/wasi-common/cap-std-sync/src/net.rs
@@ -98,7 +98,7 @@ macro_rules! wasi_listen_write_impl {
             }
             async fn sock_accept(&self, fdflags: FdFlags) -> Result<Box<dyn WasiFile>, Error> {
                 let (stream, _) = self.0.accept()?;
-                let stream = <$stream>::from_cap_std(stream);
+                let mut stream = <$stream>::from_cap_std(stream);
                 stream.set_fdflags(fdflags).await?;
                 Ok(Box::new(stream))
             }
@@ -110,7 +110,7 @@ macro_rules! wasi_listen_write_impl {
                 let fdflags = get_fd_flags(&self.0)?;
                 Ok(fdflags)
             }
-            async fn set_fdflags(&self, fdflags: FdFlags) -> Result<(), Error> {
+            async fn set_fdflags(&mut self, fdflags: FdFlags) -> Result<(), Error> {
                 if fdflags == wasi_common::file::FdFlags::NONBLOCK {
                     self.0.set_nonblocking(true)?;
                 } else if fdflags.is_empty() {
@@ -197,7 +197,7 @@ macro_rules! wasi_stream_write_impl {
                 let fdflags = get_fd_flags(&self.0)?;
                 Ok(fdflags)
             }
-            async fn set_fdflags(&self, fdflags: FdFlags) -> Result<(), Error> {
+            async fn set_fdflags(&mut self, fdflags: FdFlags) -> Result<(), Error> {
                 if fdflags == wasi_common::file::FdFlags::NONBLOCK {
                     self.0.set_nonblocking(true)?;
                 } else if fdflags.is_empty() {
diff --git a/crates/wasi-common/src/file.rs b/crates/wasi-common/src/file.rs
index b76278373..828307255 100644
--- a/crates/wasi-common/src/file.rs
+++ b/crates/wasi-common/src/file.rs
@@ -64,7 +64,7 @@ pub trait WasiFile: Send + Sync {
         Ok(FdFlags::empty())
     }
 
-    async fn set_fdflags(&self, _flags: FdFlags) -> Result<(), Error> {
+    async fn set_fdflags(&mut self, _flags: FdFlags) -> Result<(), Error> {
         Err(Error::badf())
     }
 
@@ -217,11 +217,15 @@ pub struct Filestat {
 
 pub(crate) trait TableFileExt {
     fn get_file(&self, fd: u32) -> Result<Arc<FileEntry>, Error>;
+    fn get_file_mut(&mut self, fd: u32) -> Result<&mut FileEntry, Error>;
 }
 impl TableFileExt for crate::table::Table {
     fn get_file(&self, fd: u32) -> Result<Arc<FileEntry>, Error> {
         self.get(fd)
     }
+    fn get_file_mut(&mut self, fd: u32) -> Result<&mut FileEntry, Error> {
+        self.get_mut(fd)
+    }
 }
 
 pub(crate) struct FileEntry {
@@ -272,6 +276,7 @@ impl FileEntry {
 
 pub trait FileEntryExt {
     fn get_cap(&self, caps: FileCaps) -> Result<&dyn WasiFile, Error>;
+    fn get_cap_mut(&mut self, caps: FileCaps) -> Result<&mut dyn WasiFile, Error>;
 }
 
 impl FileEntryExt for FileEntry {
@@ -279,6 +284,10 @@ impl FileEntryExt for FileEntry {
         self.capable_of(caps)?;
         Ok(&*self.file)
     }
+    fn get_cap_mut(&mut self, caps: FileCaps) -> Result<&mut dyn WasiFile, Error> {
+        self.capable_of(caps)?;
+        Ok(&mut *self.file)
+    }
 }
 
 bitflags! {
diff --git a/crates/wasi-common/src/snapshots/preview_1.rs b/crates/wasi-common/src/snapshots/preview_1.rs
index 4776cb10c..ab2af7d14 100644
--- a/crates/wasi-common/src/snapshots/preview_1.rs
+++ b/crates/wasi-common/src/snapshots/preview_1.rs
@@ -185,9 +185,9 @@ impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiCtx {
         fd: types::Fd,
         flags: types::Fdflags,
     ) -> Result<(), Error> {
-        self.table()
-            .get_file(u32::from(fd))?
-            .get_cap(FileCaps::FDSTAT_SET_FLAGS)?
+        self.table
+            .get_file_mut(u32::from(fd))?
+            .get_cap_mut(FileCaps::FDSTAT_SET_FLAGS)?
             .set_fdflags(FdFlags::from(flags))
             .await
     }
diff --git a/crates/wasi-common/src/table.rs b/crates/wasi-common/src/table.rs
index a92fefff6..7f7939a2f 100644
--- a/crates/wasi-common/src/table.rs
+++ b/crates/wasi-common/src/table.rs
@@ -76,6 +76,21 @@ impl Table {
         }
     }
 
+    /// TODO
+    pub fn get_mut<T: Any>(&mut self, key: u32) -> Result<&mut T, Error> {
+        let entry = match self.0.get_mut().unwrap().map.get_mut(&key) {
+            Some(entry) => entry,
+            None => return Err(Error::badf().context("key not in table")),
+        };
+        let entry = match Arc::get_mut(entry) {
+            Some(entry) => entry,
+            None => return Err(Error::badf().context("cannot mutably borrow shared file")),
+        };
+        entry
+            .downcast_mut::<T>()
+            .ok_or_else(|| Error::badf().context("element is a different type"))
+    }
+
     /// Remove a resource at a given index from the table. Returns the resource
     /// if it was present.
     pub fn delete<T: Any + Send + Sync>(&self, key: u32) -> Option<Arc<T>> {
```

</details>

**Comment by alexcrichton** [comment]
This has now all landed in-spirit at least with https://github.com/bytecodealliance/wasmtime/pull/5484 so I'm going to close this. Thanks for the work here though @haraldh!
